Failing to create decorator bean on interface with overloaded methods

### Describe the bug

I am trying to create a simple CDI decorator on an interface with multiple methods. Some of the interface methods are overloaded.

We have an interface like:
```java
public interface FooService {
    Foo findById(int id);
    Foo findById(String idAsString); // overload with a String argument
}
```  
And a decorator like:
```java
@Priority(100)
@Decorator
public class SomeFooServiceDecorator implements FooService {
   @Inject @Any @Delegate FooService delegate;
   // ... implement interface methods ... //
}
``` 

And finally, a bean class like:
```java
@ApplicationScoped
public class DefaultFooService implements FooService {
   // ... implement interface methods ... //
}
```


### Expected behavior

Expected a FooService bean to be created as a `SomeFooServiceDecorator` wrapping (decorating) real service bean of type `DefaultFooService`.

### Actual behavior

Dependency injection fails with `NullPointerException` errors. The exception trace contains no relevant information on the cause of the error.

A typical exception trace:
```
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.13.4.Final:build (default) on project hello-decorator: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: java.lang.IllegalStateException: java.util.concurrent.ExecutionException: java.lang.NullPointerException
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:918)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:829)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException
[ERROR] 	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
[ERROR] 	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
[ERROR] 	at io.quarkus.arc.processor.BeanProcessor.generateResources(BeanProcessor.java:322)
[ERROR] 	at io.quarkus.arc.deployment.ArcProcessor.generateResources(ArcProcessor.java:565)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
[ERROR] 	... 6 more
[ERROR] Caused by: java.lang.NullPointerException
[ERROR] 	at io.quarkus.gizmo.BytecodeCreatorImpl.allocateLocalVariables(BytecodeCreatorImpl.java:1202)
[ERROR] 	at io.quarkus.gizmo.MethodCreatorImpl.write(MethodCreatorImpl.java:122)
[ERROR] 	at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:202)
[ERROR] 	at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:225)
[ERROR] 	at io.quarkus.arc.processor.SubclassGenerator.generate(SubclassGenerator.java:127)
[ERROR] 	at io.quarkus.arc.processor.BeanProcessor$4$2.call(BeanProcessor.java:289)
[ERROR] 	at io.quarkus.arc.processor.BeanProcessor$4$2.call(BeanProcessor.java:286)
[ERROR] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[ERROR] 	... 5 more
```

### How to Reproduce?

A reproducer project is available at: 
https://github.com/drmalex07/reproducer-for-arc-bug-on-decorator

Just run Maven build (`mvn package -DskipTests`)

### Output of `uname -a` or `ver`

Linux optiplex 4.9.0-15-amd64 #1 SMP Debian 4.9.258-1 (2021-03-08) x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.14" 2022-01-18 OpenJDK Runtime Environment GraalVM CE 22.0.0.2 (build 11.0.14+9-jvmci-22.0-b05) OpenJDK 64-Bit Server VM GraalVM CE 22.0.0.2 (build 11.0.14+9-jvmci-22.0-b05, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.13.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.5 (3599d3414f046de2324203b78ddcf9b5e4388aa0) Maven home: /home/malex/opt/apache-maven-3.8.5 Java version: 11.0.14, vendor: GraalVM Community, runtime: /home/malex/opt/graalvm-ce-java11-22.0.0.2 Default locale: en_US, platform encoding: UTF-8 OS name: "linux", version: "4.9.0-15-amd64", arch: "amd64", family: "unix"

### Additional information

If we rename the overloaded method (e.g `findById` to `findById1`), everything works fine and Arc manages to resolve dependencies. But this is not a practical solution as it is very common for interfaces to contain overloaded methods.

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
@@ -649,7 +649,7 @@ public class BeanInfo implements InjectionTargetInfo {
         }
     }
 
-    List<DecoratorInfo> findMatchingDecorators(MethodInfo method, List<DecoratorInfo> decorators) {
+    private List<DecoratorInfo> findMatchingDecorators(MethodInfo method, List<DecoratorInfo> decorators) {
         List<Type> methodParams = method.parameterTypes();
         List<DecoratorInfo> matching = new ArrayList<>(decorators.size());
         for (DecoratorInfo decorator : decorators) {
@@ -661,6 +661,10 @@ public class BeanInfo implements InjectionTargetInfo {
                     throw new DefinitionException(
                             "The class of the decorated type " + decoratedType + " was not found in the index");
                 }
+
+                Map<String, Type> resolvedTypeParameters = Types.resolveDecoratedTypeParams(decoratedTypeClass,
+                        decorator);
+
                 for (MethodInfo decoratedMethod : decoratedTypeClass.methods()) {
                     if (!method.name().equals(decoratedMethod.name())) {
                         continue;
@@ -671,10 +675,12 @@ public class BeanInfo implements InjectionTargetInfo {
                     }
                     // Match the resolved parameter types
                     boolean matches = true;
-                    decoratedMethodParams = Types.getResolvedParameters(decoratedTypeClass, method,
+                    decoratedMethodParams = Types.getResolvedParameters(decoratedTypeClass, resolvedTypeParameters,
+                            decoratedMethod,
                             beanDeployment.getBeanArchiveIndex());
                     for (int i = 0; i < methodParams.size(); i++) {
-                        if (!methodParams.get(i).equals(decoratedMethodParams.get(i))) {
+                        if (!beanDeployment.getDelegateInjectionPointResolver().matches(decoratedMethodParams.get(i),
+                                methodParams.get(i))) {
                             matches = false;
                         }
                     }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DecoratorGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DecoratorGenerator.java
@@ -183,24 +183,14 @@ public class DecoratorGenerator extends BeanGenerator {
             ClassInfo decoratedTypeClass = index
                     .getClassByName(decoratedType.name());
             if (decoratedTypeClass == null) {
-                throw new IllegalStateException("TODO");
+                throw new IllegalStateException("Decorated type not found in the bean archive index: " + decoratedType);
             }
 
             // A decorated type can declare type parameters
             // For example Converter<String> should result in a T -> String mapping
             List<TypeVariable> typeParameters = decoratedTypeClass.typeParameters();
-            Map<String, org.jboss.jandex.Type> resolvedTypeParameters = Collections.emptyMap();
-            if (!typeParameters.isEmpty()) {
-                resolvedTypeParameters = new HashMap<>();
-                // The delegate type can be used to infer the parameter types
-                org.jboss.jandex.Type type = decorator.getDelegateType();
-                if (type.kind() == Kind.PARAMETERIZED_TYPE) {
-                    List<org.jboss.jandex.Type> typeArguments = type.asParameterizedType().arguments();
-                    for (int i = 0; i < typeParameters.size(); i++) {
-                        resolvedTypeParameters.put(typeParameters.get(i).identifier(), typeArguments.get(i));
-                    }
-                }
-            }
+            Map<String, org.jboss.jandex.Type> resolvedTypeParameters = Types.resolveDecoratedTypeParams(decoratedTypeClass,
+                    decorator);
 
             for (MethodInfo method : decoratedTypeClass.methods()) {
                 if (Methods.skipForDelegateSubclass(method)) {

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
@@ -675,7 +675,11 @@ public class SubclassGenerator extends AbstractGenerator {
             paramHandles[paramIdx++] = subclassConstructor.getThis();
         }
         for (DecoratorInfo decoratorParameter : decoratorParameters) {
-            paramHandles[paramIdx++] = decoratorToResultHandle.get(decoratorParameter.getIdentifier());
+            ResultHandle decoratorHandle = decoratorToResultHandle.get(decoratorParameter.getIdentifier());
+            if (decoratorHandle == null) {
+                throw new IllegalStateException("Decorator handle must not be null");
+            }
+            paramHandles[paramIdx++] = decoratorHandle;
         }
         ResultHandle delegateSubclassInstance = subclassConstructor.newInstance(MethodDescriptor.ofConstructor(
                 delegateSubclass.getClassName(), constructorParameterTypes.toArray(new String[0])), paramHandles);

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
@@ -384,8 +384,23 @@ public final class Types {
         return restrictBeanTypes(types, beanDeployment.getAnnotations(classInfo), beanDeployment.getBeanArchiveIndex());
     }
 
-    static List<Type> getResolvedParameters(ClassInfo classInfo, MethodInfo method, IndexView index) {
-        return getResolvedParameters(classInfo, Collections.emptyMap(), method, index);
+    static Map<String, Type> resolveDecoratedTypeParams(ClassInfo decoratedTypeClass, DecoratorInfo decorator) {
+        // A decorated type can declare type parameters
+        // For example Converter<String> should result in a T -> String mapping
+        List<TypeVariable> typeParameters = decoratedTypeClass.typeParameters();
+        Map<String, org.jboss.jandex.Type> resolvedTypeParameters = Collections.emptyMap();
+        if (!typeParameters.isEmpty()) {
+            resolvedTypeParameters = new HashMap<>();
+            // The delegate type can be used to infer the parameter types
+            org.jboss.jandex.Type type = decorator.getDelegateType();
+            if (type.kind() == Kind.PARAMETERIZED_TYPE) {
+                List<org.jboss.jandex.Type> typeArguments = type.asParameterizedType().arguments();
+                for (int i = 0; i < typeParameters.size(); i++) {
+                    resolvedTypeParameters.put(typeParameters.get(i).identifier(), typeArguments.get(i));
+                }
+            }
+        }
+        return resolvedTypeParameters;
     }
 
     static List<Type> getResolvedParameters(ClassInfo classInfo, Map<String, Type> resolvedMap,
@@ -588,7 +603,7 @@ public final class Types {
                         resolvedTypeParameters, index);
                 Type[] typeParams = new Type[typeParameters.size()];
                 for (int i = 0; i < typeParameters.size(); i++) {
-                    typeParams[i] = resolveTypeParam(arguments.get(i), resolvedMap, index);
+                    typeParams[i] = resolveTypeParam(typeParameters.get(i), resolvedMap, index);
                 }
                 return ParameterizedType.create(parameterizedType.name(), typeParams, null);
             }

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/SimpleDecoratorOverloadingTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/SimpleDecoratorOverloadingTest.java
@@ -1,0 +1,75 @@
+package io.quarkus.arc.test.decorators;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import javax.annotation.Priority;
+import javax.decorator.Decorator;
+import javax.decorator.Delegate;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Dependent;
+import javax.inject.Inject;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.test.ArcTestContainer;
+
+public class SimpleDecoratorOverloadingTest {
+
+    @RegisterExtension
+    public ArcTestContainer container = new ArcTestContainer(Converter.class, SimpleConverter.class,
+            ConverterDecorator.class);
+
+    @Test
+    public void testDecoration() {
+        SimpleConverter converter = Arc.container().instance(SimpleConverter.class).get();
+        assertEquals("HOLA!", converter.convert(" holA!"));
+        assertEquals(42, converter.convert(42));
+    }
+
+    interface Converter {
+
+        int convert(int value);
+
+        String convert(String value);
+
+    }
+
+    @ApplicationScoped
+    static class SimpleConverter implements Converter {
+
+        @Override
+        public String convert(String value) {
+            return value.toUpperCase();
+        }
+
+        @Override
+        public int convert(int value) {
+            return -1 * value;
+        }
+
+    }
+
+    @Dependent
+    @Priority(1)
+    @Decorator
+    static class ConverterDecorator implements Converter {
+
+        @Inject
+        @Delegate
+        Converter delegate;
+
+        @Override
+        public String convert(String value) {
+            return delegate.convert(value.trim());
+        }
+
+        @Override
+        public int convert(int value) {
+            return -1 * delegate.convert(value);
+        }
+
+    }
+
+}

